### PR TITLE
Add support for multi-step tours

### PR DESCRIPTION
**Pull request by vikaspotluri123**
refs gradebook/client#310

I'd like to get the review approval early because this includes a (relatively simple) migration

⚠ This requires a parallel merge + release of the client! ⚠ 
⚠ This cannot be merged until client is ready!            ⚠ 

**Meta**
Merge: Rebase
Reviewers: 1 + 1

### CODE REVIEW

File: lib/services/validation/user-settings.js [hunk 1]
```diff
@@ -2,11 +2,27 @@
...
+ * @description Determine if {value} is a valid number between {lowerBounds} and {upperBounds} (inclusive)
+ * @param {number} lowerBounds
+ * @param {number} upperBounds
+ * @returns {(value: any) => boolean}
+ */
+const isBetween = (lowerBounds, upperBounds) =>
+	value => !isNaN(value) && value >= lowerBounds && value <= upperBounds;
+
+// Allows for 24 states
+const LARGEST_TOUR_STEP = 0b111111111111111111111111111111111111;
```
Comment (joshcosta) [inline lib/services/validation/user-settings.js:21]: I'm counting 36

File: lib/database/migrations/000006-user-tour-numeric-value.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	// @TODO: get the proper number
+	return knex.raw(makeQuery(Number(100000).toString()));
+}
+
+/**
+ * @param {import('knex').Knex} knex
+ */
+async function down(knex) {
+	const isMysql = knex.client.config.client !== 'sqlite3';
+	return knex.raw(makeQuery(isMysql ? 'true' : 'json("true")'));
```
Comment (joshcosta) [inline lib/database/migrations/000006-user-tour-numeric-value.js:20]: Could check if tour is all zeroes and store false in that case

File: lib/services/validation/user-settings.js [hunk 3]
```diff
@@ -2,11 +2,27 @@
...
 
+/**
+ * @description Determine if {value} is a valid number between {lowerBounds} and {upperBounds} (inclusive)
+ * @param {number} lowerBounds
+ * @param {number} upperBounds
+ * @returns {(value: any) => boolean}
+ */
+const isBetween = (lowerBounds, upperBounds) =>
+	value => !isNaN(value) && value >= lowerBounds && value <= upperBounds;
+
```
Comment (ramsays) [inline lib/services/validation/user-settings.js:19]: Possible utils we could generalize? I guess only if needed

File: lib/controllers/user.js [hunk 4]
```diff
@@ -206,10 +191,12 @@ module.exports = {
...
-			// @todo: this should throw an error if validation fails
-			const course = convertHashToCourse((request.query.import || '').toString()) || makeDemoCourse();
-			course.user = user.id;
-			await api.course.create(course, txn, database);
+			if (request.query.import) {
+				// @todo: this should throw an error if validation fails
+				const course = convertHashToCourse(request.query.import.toString());
+				course.user = user.id;
+				await api.course.create(course, txn, database);
+			}
```
Comment (ramsays) [inline lib/controllers/user.js:199]: Nice, glad we don't create demo courses anymore! 👍

File: lib/database/migrations/000006-user-tour-numeric-value.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// @ts-check
+
+/** @param {string} value */
+const makeQuery = value =>
+	`UPDATE users SET settings = JSON_REPLACE(settings, "$.tour", ${value}) WHERE JSON_EXTRACT(settings, "$.tour");`;
```
Comment (ramsays) [inline lib/database/migrations/000006-user-tour-numeric-value.js:5]: This is slick! You learn something new everyday 👀

File: lib/database/migrations/000006-user-tour-numeric-value.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * @param {import('knex').Knex} knex
+ */
+async function up(knex) {
+	// @TODO: get the proper number
+	return knex.raw(makeQuery(Number(100000).toString()));
```
Comment (ramsays) [inline lib/database/migrations/000006-user-tour-numeric-value.js:12]: What exactly is this doing? I'm not sure I understand.

**Review by ramsays — APPROVED** [review]

**Review by joshcosta — APPROVED** [review]
31 is right for 4 steps + 1 initial

**Pull request merged**